### PR TITLE
cmd, core, eth, miner, params, tests: finalize the DAO fork

### PR DESCRIPTION
**Pull request by karalabe**
*Note: This PR contains all the pieces of the DAO hard-fork, each detailed below. The remaining PRs tagged `dao-fork-part` may be disregarded, they were meant to aid review.*

*Note 2: After this PR is merged, we need to prep a merge into the `release/1.4` branch and inherently `master`, which needs a tiny amount of work due to the databse reogs already on `develop`.*

---

##### *cmd, core, eth, params: implement flags to control dao fork blocks* #2788

This PR implements setting the `--support-dao-fork` and `--oppose-dao-fork` flags. As of now they only modify a single database entry specifying whether the current default behavior should change.

 * If no dao fork number is set, `--support-dao-fork` will set it to either mainnet/testnet number
 * If a dao fork number is set, `--support-dao-fork` will leave it at its present value
 * If a dao fork number is set, `--oppose-dao-fork` will remove it

*Caveat: setting the dao number from a custom genesis file, removing it via `--oppose-dao-fork` and reenabling it via `--support-dao-fork` will result in the mainnet/testnet number being set, the original private number being lost. Don't use this willy nilly on private networks; rather use `geth init genesis.json` for those scenarios.*

This PR further drops support to the deprecated `--genesis` flag. This is important because the `--support-dao-fork` and `--oppose-dao-fork` flags need to know the previous state/number in the database to ensure it's not accidentally overwritten! We do not know this with the `--genesis` flag as the user might supply something we can only interpret much later in the code. *Note, for main and test network we know in advance what their genesis file would contain, so that's not an issue*.

##### *core: gracefully handle missing homestead block config* #2790 

Our nodes didn't handle an interesting corner case until now: **having** a chain configuration (in the genesis.json), but **not having** a `homesteadBlock` set in it (i.e. `{..., "config": {}}`). Since our configs until now only featured the homestead block number, either it (block number) was present or the entire config section was missing (resulting the the defaults being used).

With the DAO fork being worked on and requiring an additional config section, it becomes entirely possible to have the DAO fork number set, but the homestead number not set. This PR ensures that it the homestead block number is missing, it's assumed not forked.

*Note, there's almost no end user use case to set the DAO hard fork in a private network (so skipping the entire config section or only setting the homestead block is still entirely workable), but it is needed to properly test it n automated ways so better to fix it up properly now.*

##### *cmd/geth, miner, params: special extradata for DAO fork start* #2791 

This PR modifies the miner so that if the DAO hard-fork is enabled, the fork block and `10` consecutive blocks all have `0x64616f2d686172642d666f726b` (hex for `dao-hard-fork`) as their extra-data.

This is important to allow header-only mode clients (fast sync algo, light client) to verify whether a header is indeed on the correct chain and not get duped into syncing a live alternative fork. It also helps full nodes to split the network between two forks and avoid cross network noise.

Consecutive blocks are needed to prevent any malicious miner from minting a non-forked block with the fork extra-data. Unless almost all non-forking miners are malicious, there will be a non-forking block in the first `10` that does not have the fork extra-data set, separating that chain from the forked chain.

This PR does not contain unit tests as out miner module wasn't written with unit tests in mind and I'd like to avoid a too invasive rewrite in such a critical time. However, this feature **is** covered by [`hive` end to end tests](https://github.com/karalabe/hive/tree/master/simulators/dao-hard-fork/extradata) with full mining.

##### *cmd, core, miner: add extradata validation to consensus rules* #2794

This PR makes the necessary consensus modifications to header validation rules so that block header `extraData` fields are enforced to be either `0x64616f2d686172642d666f726b` (hex for `dao-hard-fork`) for pro-forkers or everything **but** that for non-forkers. These rules are upheld for `10` blocks straight to protect both forkers and non-forkers against malicious actors of the opposite side.

Since non-forkers also need to know the actual DAO hard-fork block number to enforce themselves onto the opposite side of the network (i.e. a non-forker fast syncing needs to be aware of the headers it needs to check), this PR makes a further modification to the previous chain configurations so not only a single `ChainConfig.DAOForkNumber` is maintained (previously implicitly signalling pro-fork), but rather an explicit `ChainConfig.DAOForkSupport` boolean flag is added too. This way both camps can maintain the block numbers whilst also allowing both decisions.

##### *core, eth: enforce network split post DAO hard-fork* #2795

This PR is meant to enforce the network split after passing the DAO hard-fork block number. It is done by requesting the DAO fork-block's header immediately after an `eth` handshake completes. If the DAO challenge isn't completed within `15 seconds`, the connection is dropped.

To complete the DAO challenge, the remote peer must reply with the requested header (every peer will do so, it's how the `eth` protocol works). A few things can happen:

 * The returned header's extra-data conforms to the side we're on. Challenge completed.
 * The returned header's extra-data doesn't conform to our side. Challenge failed, peer dropped.
 * No header is returned, this complicates things:
   * We ourselves haven't reached the fork point, **assume** friendly (no way to challenge)
   * We've already reached the fork point, so we already have our fork block:
     * If peer's TD is smaller than this, we cannot check it's side, **assume** friendly
     * If peer's TD is higher than this, assume the packet is a reply to something else, **wait** further

##### *core, params, tests: add DAO hard-fork balance moves* #2800

This PR implements the last component of the DAO hard-fork, namely defines the list of DAO addresses the balances of which should be moved and executes the actual state transformation when the DAO hard-fork block arrives.

The PR also adds the DAO consensus tests from the upstream https://github.com/ethereum/tests repo, although as far as I know those tests are still under development, and hence are subject to change.

*Note, the fork parameters (block number and refund account) are just placeholders, whereas the DAO drain list may or may not change.*

##### *cmd, core, eth, miner, params, tests: finalize the DAO fork* #2814

The last commit for the DAO hard-fork. This finalizes the DAO child list, the refund contract address, the testnet and mainnet transition blocks as well as sets the default behavior of Geth to be pro-fork as per internal discussions.

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Thu Jul 14 08:26:15 UTC 2016_

**Comment by ebuchman** [comment]
Is there a spec for the hard fork independent of the PRs that you could link to?

Is there no plan to prevent transaction replay across forks?

**Comment by jarradh** [comment]
> sets the default behavior of Geth to be pro-fork as per internal discussions
...

**Comment by julian1** [comment]
There are 35 file changes, including modifications to the configuration/genesis, p2p networking, and EVM protocol. Given the inherent complexity, I wonder if the default should be the option with less technical risk.

**Comment by julian1** [comment]
> as well as sets the default behavior of Geth to be pro-fork as per internal discussions.


The application could randomize this boolean on first-start when no explicit flag to support or oppose the fork is present.

This would help shield developers from accusations of bias and protect Geth neutrality.

Was this possibility considered?

**Comment by jarradh** [comment]
The explicit flag for or against makes far more sense to me

**Comment by karalabe** [comment]
@julian1 Most of the change files are tests (22/35). The actual changes aren't **that** huge, but I agree that it's a non trivial change. Regarding the defaults, I believe @obscuren is prepping a post about it.

File: tests/vm_test_util.go [hunk 1]
```diff
@@ -241,7 +241,7 @@ func RunVm(state *state.StateDB, env, exec map[string]string) ([]byte, vm.Logs,
 
 	caller := state.GetOrNewStateObject(from)
 
-	vmenv := NewEnvFromMap(RuleSet{params.MainNetHomesteadBlock}, state, env, exec)
+	vmenv := NewEnvFromMap(RuleSet{params.MainNetHomesteadBlock, params.MainNetDAOForkBlock, true}, state, env, exec)
```
Comment (ileathan) [inline tests/vm_test_util.go]: .

File: cmd/ethtest/main.go [hunk 2]
```diff
@@ -74,9 +74,9 @@ func runTestWithReader(test string, r io.Reader) error {
 	var err error
 	switch strings.ToLower(test) {
 	case "bk", "block", "blocktest", "blockchaintest", "blocktests", "blockchaintests":
-		err = tests.RunBlockTestWithReader(params.MainNetHomesteadBlock, r, skipTests)
+		err = tests.RunBlockTestWithReader(params.MainNetHomesteadBlock, params.MainNetDAOForkBlock, r, skipTests)
 	case "st", "state", "statetest", "statetests":
-		rs := tests.RuleSet{HomesteadBlock: params.MainNetHomesteadBlock}
+		rs := tests.RuleSet{HomesteadBlock: params.MainNetHomesteadBlock, DAOForkBlock: params.MainNetDAOForkBlock, DAOForkSupport: true}
```
Comment (ileathan) [inline cmd/ethtest/main.go]: .

File: cmd/geth/dao_test.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+//
+// go-ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with go-ethereum. If not, see <http://www.gnu.org/licenses/>.
+
+package main
```
Comment (ileathan) [inline cmd/geth/dao_test.go]: ..

**Comment by nomebahi** [comment]
great project

**Comment by Ekhim04** [comment]
karalabe:dao-hard-finalcombo

**Comment by Ekhim04** [comment]
[](url)karalabe:dao-hard-finalcombo

**Pull request merged**